### ISSUE TITLE
Partial service --image update with start-first && parallelism on registry timeout

### ISSUE THREAD

**Issue by dejeroWilliamScott** [moby/moby#44069]
### Description

When a service definition update_config includes `order: start-first` and `parallelism: N`, and a Docker registry is temporarily unreachable (e.g. DNS failure), the service can enter a mixed deployment state.  This mixed state is undesirable, as there are containers that are part of a service that do not adhere to the service definition (e.g. mixed versions).

### (1) Create a service with the related `update_config` definition
`docker stack deploy RepStack -c my_yaml.yaml --with-registry-auth`
```
services:
  RepService:
    image: "${RepImage}"
    deploy:
      replicas: 6
      update_config:
        parallelism: 3
        delay: 5s
        order: start-first
```

### (2) Verify the service deployed successfully
`docker service ls`
```
ID             NAME                                     MODE         REPLICAS   IMAGE
xezchytsyhvz   RepStack_RepService                      replicated   6/6        RepImage:14.0.0-m318.fdc2576a8
```

### (3) Block access to the Docker registry where images are kept
`<various methods, firewall rule is easiest>`

### (4) Perform a service update to reference a new image
`docker service update RepStack_RepService --image=14.0.0-m319.ac2334a0f --with-registry-auth`

### (6) Wait for the update to fail (cannot pull image)
Note that the containers relating to the update_config start-first parellelism will continue to try to pull the image after the update has failed.

### (5) Restore access to the Docker registry
`<various methods, firewall rule is easiest>`

### (6) Observe the update_config containers pull the image
The service definition will reference the new image, but the majority of the containers will remain the old version. 
`docker service ls`
```
ID             NAME                                     MODE         REPLICAS   IMAGE
xezchytsyhvz   RepStack_RepService                      replicated   9/6        RepImage:14.0.0-m319.ac2334a0f      
```
`docker service ps RepStack_RepService | grep Running`
```
t14i228um4fb   my_node.1       RepImage:14.0.0-m319.ac2334a0f   sqa-dmz-3   Running         Running 19 hours ago
zqmezvg1mgqf    \_ my_node.1   RepImage:14.0.0-m318.fdc2576a8   sqa-dmz-3   Running         Running 43 hours ago
uc8pp8grunk5   my_node.2       RepImage:14.0.0-m318.fdc2576a8   sqa-dmz-3   Running         Running 43 hours ago
nnjvb01xiq6a   my_node.3       RepImage:14.0.0-m319.ac2334a0f   sqa-dmz-3   Running         Running 19 hours ago
l08nsda35hbs    \_ my_node.3   RepImage:14.0.0-m318.fdc2576a8   sqa-dmz-1   Running         Running 43 hours ago
j37ulmmcx56n   my_node.4       RepImage:14.0.0-m318.fdc2576a8   sqa-dmz-2   Running         Running 43 hours ago
bx7u4t215lmm   my_node.5       RepImage:14.0.0-m318.fdc2576a8   sqa-dmz-1   Running         Running 43 hours ago
k6w82uux6bmw   my_node.6       RepImage:14.0.0-m319.ac2334a0f   sqa-dmz-3   Running         Running 19 hours ago
jz7nmh0hy2b4    \_ my_node.6   RepImage:14.0.0-m318.fdc2576a8   sqa-dmz-3   Running         Running 43 hours ago
```
A quick tally shows the three containers from the update_config have joined, which meets the service definition, but the old containers persist.  All containers serve requests for the service definition, and consequently, a long-lived service may operate with "mixed" results (e.g. serve requests with non-uniform image versions).
```
6x RepImage:14.0.0-m318.fdc2576a8 
3x RepImage:14.0.0-m319.ac2334a0f 
```

### Expected behavior

Either:
 * The service definition update should fail after the parallelism containers failed to update and be rolled back to the previous definition (after a timeout)
 * The service definition update should be applied after the parallelism containers start (no update timeout)

I expect the latter to be less transparent, as there would need to be a method to view "updates in progress" (without such visibility, there is preference to the former).

### docker version

```bash
docker version
Client: Docker Engine - Community
 Version:           20.10.15
 API version:       1.41
 Go version:        go1.17.9
 Git commit:        fd82621
 Built:             Thu May  5 13:19:18 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.17
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.17.11
  Git commit:       a89b842
  Built:            Mon Jun  6 23:01:02 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.4
  GitCommit:        212e8b6fa2f44b9c21b2798135fc6fb7c53efc16
 runc:
  Version:          1.1.1
  GitCommit:        v1.1.1-0-g52de29d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.8.2-docker)
  scan: Docker Scan (Docker Inc., v0.17.0)

Server:
 Containers: 5
  Running: 5
  Paused: 0
  Stopped: 0
 Images: 5
 Server Version: 20.10.17
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: false
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: active
  NodeID: xo8y9ypxpfm8mgq5v8x4fs4n9
  Is Manager: true
  ClusterID: mjyrlhhcgkbdmsrivb75gf79d
  Managers: 3
  Nodes: 12
  Default Address Pool: 10.0.0.0/8
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 10.105.127.51
  Manager Addresses:
   10.105.127.51:2377
   10.105.127.52:2377
   10.105.127.53:2377
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 212e8b6fa2f44b9c21b2798135fc6fb7c53efc16
 runc version: v1.1.1-0-g52de29d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.4.0-1072-azure
 Operating System: Ubuntu 18.04.6 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 15.63GiB
 Name: sqa-mgr-1
 ID: IY4X:C2IK:I3UE:7KBG:KD57:Q34J:YD3Y:WMOO:HVRL:CP5Z:Q7LQ:EIWQ
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support
```

**Comment by thaJeztah** [comment]
> - The service definition update should fail after the parallelism containers failed to update and be rolled back to the previous definition (after a timeout)

Based on your example, it looks like your deploy configuration uses the default ("pause") failure action. In that case, swarm will pause updating replicas (tasks) if a failure happens (max_failure_ratio); see the compose-file reference for "v3" (the version of the compose file schema used for swarm stacks); https://docs.docker.com/compose/compose-file/compose-file-v3/#update_config

> #### rollback_config
> 
> > Added in [version 3.7](compose-versioning.md#version-37) file format.
> 
> Configures how the service should be rollbacked in case of a failing
> update.
> 
> - `parallelism`: The number of containers to rollback at a time. If set to 0, all containers rollback simultaneously.
> - `delay`: The time to wait between each container group's rollback (default 0s).
> - `failure_action`: What to do if a rollback fails. One of `continue` or `pause` (default `pause`)
> - `monitor`: Duration after each task update to monitor for failure `(ns|us|ms|s|m|h)` (default 5s) **Note**: Setting to 0 will use the default 5s.
> - `max_failure_ratio`: Failure rate to tolerate during a rollback (default 0).
> - `order`: Order of operations during rollbacks. One of `stop-first` (old task is stopped before starting new one), or `start-first` (new task is started first, and the running tasks briefly overlap) (default `stop-first`).
> 
> #### update_config
> 
> Configures how the service should be updated. Useful for configuring rolling
> updates.
> 
> - `parallelism`: The number of containers to update at a time.
> - `delay`: The time to wait between updating a group of containers.
> - `failure_action`: What to do if an update fails. One of `continue`, `rollback`, or `pause`
>   (default: `pause`).
> - `monitor`: Duration after each task update to monitor for failure `(ns|us|ms|s|m|h)` (default 5s) **Note**: Setting to 0 will use the default 5s.
> - `max_failure_ratio`: Failure rate to tolerate during an update.
> - `order`: Order of operations during updates. One of `stop-first` (old task is stopped before starting new one), or `start-first` (new task is started first, and the running tasks briefly overlap) (default `stop-first`) **Note**: Only supported for v3.4 and higher.

As to your second bullet:

> - The service definition update should be applied after the parallelism containers start (no update timeout) 

That approach won't be possible due to how swarm services (and other distributed systems using reconciliation) works. SwarmKit (which manges swarm services) use a raft store, which stores the _desired_ state. When you update a service definition and the Swarm managers accept the change, that configuration is the desired state. This desired state is distributed to the worker nodes where the instances are scheduled, after which the worker nodes deploy the actual instance. Failures may happen during that deploy, and depending on the deploy policy configuration described above, the cluster will either "continue" trying to deploy (until actual state reconciles with desired state), "pause" deploy (which can result in a partial deployment), or "rollback" (revert to the previous state).


Based on the above, I think this works as expected, but happy to hear if it doesn't 👍

**Comment by dejeroWilliamScott** [comment]
Thank you for the response @thaJeztah!  Your description of swarm services && SwarmKit confirms my understanding and expectation(s).

~If my understanding is correct, is it possible that the rollback action failed and that resulted in the 9/6 split (e.g. the nodes were unable to contact the registry for verification of the old image version)?  If so, would it make sense to "trust" the local old image on rollback in this case?  There may be an edgecase I'm not thinking of, but I'd like to assume that the local image is a more accurate record of the previous deployment than a remote copy.~
Nevermind all of this, the default action is `pause` (not `rollback`).

I'm trying to rationalize what could result in a deployed container version gap several days after the deployment.  To the best of my understanding, once the registry is available, the service should continue with the update (post pause).  I'd be happy to perform some tests, but I'm uncertain what exploratory work should be done.
